contracts/fuses/universal_token_swapper/UniversalTokenSwapperEthFuse.sol

```diff
@@ -0,0 +1,148 @@
+struct UniversalTokenSwapperEthEnterData {
+    address tokenIn;
+    address tokenOut;
+    uint256 amountIn;
+    UniversalTokenSwapperEthData data;
+}
+
+struct Balances {
+    uint256 tokenInBalanceBefore;
+    uint256 tokenOutBalanceBefore;
+    uint256 tokenInBalanceAfter;
+    uint256 tokenOutBalanceAfter;
+}
+
+/// @title This contract is designed to execute every swap operation and check the slippage on any DEX.
+contract UniversalTokenSwapperEthFuse is IFuseCommon {
+    using SafeERC20 for ERC20;
+
+    event UniversalTokenSwapperFuseEnter(address version, address asset, uint256 amount);
+
+    error UniversalTokenSwapperFuseUnsupportedAsset(address asset);
+    error UniversalTokenSwapperFuseSlippageFail();
+
+    address public immutable VERSION;
+    uint256 public immutable MARKET_ID;
+    address payable public immutable EXECUTOR;
+    /// @dev slippageReverse in WAD decimals, 1e18 - slippage;
+    uint256 public immutable SLIPPAGE_REVERSE;
+    uint256 private constant _ONE = 1e18;
+
+    constructor(uint256 marketId_, address executor_, uint256 slippageReverse_) {
+        VERSION = address(this);
+        MARKET_ID = marketId_;
+        EXECUTOR = payable(executor_);
+        if (slippageReverse_ > _ONE) {
+            revert UniversalTokenSwapperFuseSlippageFail();
+        }
+        SLIPPAGE_REVERSE = _ONE - slippageReverse_;
+    }
+
+    function enter(UniversalTokenSwapperEthEnterData calldata data_) external {
+        if (!PlasmaVaultConfigLib.isSubstrateAsAssetGranted(MARKET_ID, data_.tokenIn)) {
+            revert UniversalTokenSwapperFuseUnsupportedAsset(data_.tokenIn);
+        }
+        if (!PlasmaVaultConfigLib.isSubstrateAsAssetGranted(MARKET_ID, data_.tokenOut)) {
+            revert UniversalTokenSwapperFuseUnsupportedAsset(data_.tokenOut);
+        }
+
+        uint256 dexsLength = data_.data.targets.length;
+
+        for (uint256 i; i < dexsLength; ++i) {
+            if (!PlasmaVaultConfigLib.isSubstrateAsAssetGranted(MARKET_ID, data_.data.targets[i])) {
+                revert UniversalTokenSwapperFuseUnsupportedAsset(data_.data.targets[i]);
+            }
+        }
+
+        address plasmaVault = address(this);
+
+        Balances memory balances = Balances({
+            tokenInBalanceBefore: ERC20(data_.tokenIn).balanceOf(plasmaVault),
+            tokenOutBalanceBefore: ERC20(data_.tokenOut).balanceOf(plasmaVault),
+            tokenInBalanceAfter: 0,
+            tokenOutBalanceAfter: 0
+        });
+
+        if (data_.amountIn == 0) {
+            return;
+        }
+
+        ERC20(data_.tokenIn).safeTransfer(EXECUTOR, data_.amountIn);
+
+        SwapExecutorEth(EXECUTOR).execute(
+            SwapExecutorEthData({
+                tokenIn: data_.tokenIn,
+                tokenOut: data_.tokenOut,
+                targets: data_.data.targets,
+                callData: data_.data.callData,
+                ethAmounts: data_.data.ethAmounts,
+                dustToCheck: data_.data.dustToCheck
```

Comment:
lack of validation that tokens used in dustToCheck list are in substrates set

Reply:
And why do we need such validation? 
This is not in the context of plasmavault, so the call should not generate any risk. 
If the appropriate amount does not return to the plasma vault, there will be a revert.